_posts/2018-07-17-its-a-lot-more-snappy.md

```diff
@@ -0,0 +1,174 @@
+
+When it came to inserting the data into a table in IndexedDB, we found that the native javascript/browser IndexedDB API was cumbersome and unintuitive. We found a wrapper called `Dexie` on NPM (Pete is particularly fond of the name of this module). Dexie also allowed us to perform bulk  put operations on our IndexedDB tables by opting out of listening to onSuccess callbacks for each put operation.
+
+_the magic runes_
+
+```typescript
+class PokemonDB extends Dexie {
+
+    private readonly schemaDefinition = 'pokemonId';
+    private readonly versionNumber = 1;
+
+    constructor() {
+        super('PokemonDB');
+        const stores = {};
+        for (const location of locations) {
+            stores[location] = this.schemaDefinition;
+        }
+        this.version(this.versionNumber).stores(stores);
+    }
+}
+
+class PokemonUpdater {
+
+    private db: PokemonDB;
+    private table: Dexie.Table<Pokemon, number>;
+    private pokemon: Pokemon[];
+    private readonly partitionSize = 5000;
+
+    constructor(location: string, pokemon: Floc[]) {
+        this.db = new PokemonDB();
+        this.table = this.db.table(location);
+        this.pokemon = pokemon;
+    }
+
+    public async updatePokemon(): Promise<void> {
+        const pokemonLength = this.pokemon.length;
+        let partitionStart = 0;
+        let partitionEnd = 0;
+
+        await this.db.transaction('rw', this.table, async (): Promise<void> => {
+            while (partitionStart < pokemonLength) {
+
+                const sizeDifference = pokemonLength - partitionEnd;
+
+                if (sizeDifference < this.partitionSize) {
+                    partitionEnd += sizeDifference;
+                } else {
+                    partitionEnd += this.partitionSize;
+                }
+
+                const partition = this.pokemon.slice(partitionStart, partitionEnd);
+                await this.table.bulkPut(partition);
+                partitionStart = partitionEnd;
+            }
+        });
+    }
+}
+```
+
+Now that we had implemented this pattern we noticed that the initial download of the data was taking much longer than it had previously, as we were still downloading the dataset in the same way, performing a mapping operation on the dataset before storing it in our redux store and then inserting all of our newly obtained Pokemon into our Pokemon DB. If only there existed a method, which could enable us to parallelise these operations.
+
+It was at this point our fearless tech lead, William Sia(TM) proposed using a web worker for that exact purpose. This was new and very exciting. Of course we wanted to use a web worker!
+
+I happened to come across a module called `worker-loader` which seamlessly integrated with our current build and packaging eco-system. In fact, worker-loader almost made it inconceivable that we waited such a long time to implement our first web worker. We didn’t even need to add any additional config to our web pack config files.
```

Comment:
Again - don't make it sound random.  You researched, analysed and chose...

But you don't need to say all that...

``` `worker-loader` seamlessly integrated with our current build and packaging eco-system.```